service/topology_coordinator.cc

```diff
@@ -1300,6 +1335,41 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                                 .build());
                     }
                     break;
+                case locator::tablet_transition_stage::repair: {
+                    auto& tinfo = tmap.get_tablet_info(gid.tablet);
+                    auto time = tinfo.tablet_task_info.sched_time;
+                    if (action_failed(tablet_state.repair)) {
+                        if (do_barrier()) {
+                            _tablets.erase(gid);
+                            updates.emplace_back(get_mutation_builder()
+                                    .del_transition(last_token)
+                                    .build());
+                        }
+                        break;
+                    }
+                    if (advance_in_background(gid, tablet_state.repair, "repair", [&] () -> future<> {
+                        auto primary = tmap.get_primary_replica(gid.tablet);
+                        auto dst = primary.host;
+                        auto tablet = gid;
+                        auto sched_time = time;
+                        rtlogger.info("Initiating tablet repair host={} tablet={}", dst, gid);
+                        co_await ser::storage_service_rpc_verbs::send_tablet_repair(&_messaging,
```

Comment:
Unlike other migration tasks, repair has only one stage to execute. I see we do not send a barrier in stream stage. 

```
rtlogger.info("Initiating tablet streaming ({}) of {} to {}", trinfo.transition, gid, *trinfo.pending_replica);
                        auto dst = trinfo.pending_replica->host;
                        return ser::storage_service_rpc_verbs::send_tablet_stream_data(&_messaging,
                                   netw::msg_addr(id2ip(dst)), _as, raft::server_id(dst.uuid()), gid);

```

In what case we would run a second repair for a given tablet  when the previous one is not finished yet? In coordinator fail and hand over case, the repair rpc will be joined with previous repair.

Reply:
In case of regular migration, there will be a barrier after session is cleared and before the next migration can start. Because we have stages before transition is finished (success or rollback) which will do the barrier (e.g. `end_migration` / `revert_migration`), and that will execute before the next migration starts.

> In what case we would run a second repair for a given tablet when the previous one is not finished yet?

If coordinator detects that repair RPC failed, it will proceed to null-out the transition. That clears the session, but without the barrier, there is no ordering of that and the actual end of repair processes on replicas. As soon as the transition is null, another repair may start and issue the RPC. That PRC may not join with the old repair, which may already failed, but left some processes running on other nodes. To have guarantee that no process of nulled session is running anywhere, you have to execute a barrier after nulling the session and before sending a new repair RPC.

Reply:
So since locator::tablet_transition_stage::repair is single stage, we have no other stages to send the barriers. If we always send a barrier before the repair rpc, it is guaranteed. 

How about this? Is it ok to call do_barrier inside another advance_in_background?

```
                case locator::tablet_transition_stage::repair: {
                    auto& tinfo = tmap.get_tablet_info(gid.tablet);
                    auto time = tinfo.tablet_task_info.sched_time;
                    if (action_failed(tablet_state.repair)) {
                        if (do_barrier()) {
                            _tablets.erase(gid);
                            updates.emplace_back(get_mutation_builder()
                                    .del_transition(last_token)
                                    .build());
                        }
                        break;
                    }
                    if (advance_in_background(gid, tablet_state.repair, "repair", [&] () -> future<> {
                        auto primary = tmap.get_primary_replica(gid.tablet);
                        auto dst = primary.host;
                        auto tablet = gid;
                        auto sched_time = time;
                        if (do_barrier()) {
                            rtlogger.info("Initiating tablet repair host={} tablet={}", dst, gid);
                            co_await ser::storage_service_rpc_verbs::send_tablet_repair(&_messaging,
                                    netw::msg_addr(id2ip(dst)), _as, raft::server_id(dst.uuid()), gid);
                            auto duration = std::chrono::duration<float>(db_clock::now() - sched_time);
                            rtlogger.info("Finished tablet repair host={} tablet={} duration={}", dst, tablet, duration);
                        }
                    })) {
                        if (do_barrier()) {
                            _tablets.erase(gid);
                            updates.emplace_back(get_mutation_builder()
                                .del_transition(last_token)
                                .del_tablet_task_info(last_token)
                                .set_repair_time(last_token, time)
                                .build());
                        }

                    }
                }
                    break;
            }
```

Reply:
I think it's ok, but I would avoid nesting advance_in_background(), which do_barrier() also calls, just in case they would share some things in the future. 

And have this instead:

```
  if (do_barrier()) {
      if (advance_in_background(gid, tablet_state.repair, "repair", [&] () -> future<> {
        ...
      }
  }
```

Reply:
OK. Now it looks like below. Please take another look for all the barrier usages. 

```
                case locator::tablet_transition_stage::repair: {
                    auto& tinfo = tmap.get_tablet_info(gid.tablet);
                    auto time = tinfo.tablet_task_info.sched_time;
                    if (action_failed(tablet_state.repair)) {
                        if (do_barrier()) {
                            _tablets.erase(gid);
                            updates.emplace_back(get_mutation_builder()
                                    .del_transition(last_token)
                                    .build());
                        }
                        break;
                    }
                    if (do_barrier()) {
                        if (advance_in_background(gid, tablet_state.repair, "repair", [&] () -> future<> {
                            auto primary = tmap.get_primary_replica(gid.tablet);
                            auto dst = primary.host;
                            auto tablet = gid;
                            auto sched_time = time;
                            rtlogger.info("Initiating tablet repair host={} tablet={}", dst, gid);
                            co_await ser::storage_service_rpc_verbs::send_tablet_repair(&_messaging,
                                    netw::msg_addr(id2ip(dst)), _as, raft::server_id(dst.uuid()), gid);
                            auto duration = std::chrono::duration<float>(db_clock::now() - sched_time);
                            rtlogger.info("Finished tablet repair host={} tablet={} duration={}", dst, tablet, duration);
                        })) {
                            if (do_barrier()) {
                                _tablets.erase(gid);
                                updates.emplace_back(get_mutation_builder()
                                    .del_transition(last_token)
                                    .del_tablet_task_info(last_token)
                                    .set_repair_time(last_token, time)
                                    .build());
                            }

                        }
                    }
                }
```